### PR TITLE
rabbitmq: fix having neither acked nor nacked messages

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix having neigher acked nor nacked messages. If exception happens during read-write phase, messages will be nacked.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit e55b60e05c8f564738d92bfa35f68a378732f690 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59775/e55b60e05c8f564738d92bfa35f68a378732f690/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59775/e55b60e05c8f564738d92bfa35f68a378732f690/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/RabbitMQ/RabbitMQConsumer.cpp [hunk 1]
```diff
@@ -128,6 +128,32 @@ bool RabbitMQConsumer::ackMessages(const CommitInfo & commit_info)
...
+bool RabbitMQConsumer::nackMessages(const CommitInfo & commit_info)
+{
+    if (state != State::OK)
+        return false;
+
+    /// Nothing to nack.
+    if (!commit_info.delivery_tag || commit_info.delivery_tag <= last_commited_delivery_tag)
+        return false;
+
+    if (consumer_channel->reject(commit_info.delivery_tag, AMQP::multiple))
```
Comment (jkartseva) [inline src/Storages/RabbitMQ/RabbitMQConsumer.cpp:140]: By briefly searching for the AMQP API, it distinguishes between `nack` and `reject`: https://www.rabbitmq.com/nack.html

> Unfortunately, `basic.reject` provides no support for negatively acknowledging messages in bulk.
To solve this, RabbitMQ supports the `basic.nack` method that provides all the functionality of basic.reject whilst also allowing for bulk processing of messages.
>
The API also provides a `nack` flag:
https://github.com/CopernicaMarketingSoftware/AMQP-CPP/blob/master/include/amqpcpp/flags.h#L31

I wonder if it's possible to use the AMQP API? Probably that's beyond the scope of the fix.

File: tests/integration/test_storage_rabbitmq/test.py [hunk 2]
```diff
@@ -3549,3 +3549,6 @@ def test_attach_broken_table(rabbitmq_cluster):
     assert "CANNOT_CONNECT_RABBITMQ" in error
     error = instance.query_and_get_error("INSERT INTO rabbit_queue VALUES ('test')")
     assert "CANNOT_CONNECT_RABBITMQ" in error
+
+
+# TODO: add a test
```
Comment (jkartseva) [inline tests/integration/test_storage_rabbitmq/test.py:3554]: Missing a clarification on which test should be added.

File: src/Storages/RabbitMQ/RabbitMQSource.cpp [hunk 3]
```diff
@@ -120,10 +120,20 @@ Chunk RabbitMQSource::generateImpl()
     {
         auto timeout = std::chrono::milliseconds(context->getSettingsRef().rabbitmq_max_wait_ms.totalMilliseconds());
         consumer = storage.popConsumer(timeout);
+
+        if (consumer->needChannelUpdate())
+        {
```
Comment (jkartseva) [inline src/Storages/RabbitMQ/RabbitMQSource.cpp:125]: I'm not sure if that is practical, yet `popConsumer` is designed to return `nullptr` if the waiting timeout is exceeded and no consumer has become available by that time:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/RabbitMQ/StorageRabbitMQ.cpp#L901

Should be accounted for.

**Review by jkartseva — COMMENTED** [review]
LGTM with some comments.

File: src/Storages/RabbitMQ/RabbitMQConsumer.cpp (hunk 1 above)
Comment (kssenii) [reply src/Storages/RabbitMQ/RabbitMQConsumer.cpp:140]: > Unfortunately, basic.reject provides no support for negatively acknowledging messages in bulk.

in AMQP-CPP the method `reject` has `multiple` flag so it should support bulk nacking. As far as I see in this doc https://www.rabbitmq.com/nack.html this is the only difference between `nack` and `reject`. Also near `reject` method in AMQP-CPP there is a comment "Reject or nack a message"(https://github.com/ClickHouse/AMQP-CPP/blob/00f09897ce020a84e38f87dc416af4a19c5da9ae/include/amqpcpp/channel.h#L554-L556) so should be doing both. According to https://github.com/ClickHouse/AMQP-CPP/blob/00f09897ce020a84e38f87dc416af4a19c5da9ae/src/channelimpl.cpp#L707-L713 the `reject` method does exactly what we need (sends `BasicNackFrame`). While `noAck` flag from https://github.com/ClickHouse/AMQP-CPP/blob/00f09897ce020a84e38f87dc416af4a19c5da9ae/src/flags.cpp#L27 has a comment "if set, consumed messages do not have to be acked" which means a bit different thing and is used in scope of `consume` method.
So seems using `reject` here is what we need.

File: tests/integration/test_storage_rabbitmq/test.py (hunk 2 above)
Comment (kssenii) [reply tests/integration/test_storage_rabbitmq/test.py:3554]: This is just a reminder to me to add a test before merging this PR, was not planning to leave this comment here without adding a test.

**Review by jkartseva — APPROVED** [review]

**Pull request merged**